pkg/login/social/google_oauth.go

```diff
@@ -48,3 +116,95 @@ func (s *SocialGoogle) AuthCodeURL(state string, opts ...oauth2.AuthCodeOption)
 	}
 	return s.SocialBase.AuthCodeURL(state, opts...)
 }
+
+func (s *SocialGoogle) extractFromToken(ctx context.Context, client *http.Client, token *oauth2.Token) (*googleUserData, error) {
+	s.log.Debug("Extracting user info from OAuth token")
+
+	idToken := token.Extra("id_token")
+	if idToken == nil {
+		s.log.Debug("No id_token found, defaulting to API access", "token", token)
+		return nil, nil
+	}
+
+	rawJSON, err := s.retrieveRawIDToken(idToken)
+	if err != nil {
+		s.log.Warn("Error retrieving id_token", "error", err, "token", fmt.Sprintf("%+v", idToken))
+		return nil, nil
+	}
+
+	s.log.Debug("Received id_token", "raw_json", string(rawJSON))
+
+	var data googleUserData
+	if err := json.Unmarshal(rawJSON, &data); err != nil {
+		return nil, fmt.Errorf("Error getting user info: %s", err)
+	}
+
+	return &data, nil
+}
+
+type googleGroupResp struct {
+	Memberships []struct {
+		Group    string `json:"group"`
+		GroupKey struct {
+			ID string `json:"id"`
+		} `json:"groupKey"`
+		DisplayName string `json:"displayName"`
+	} `json:"memberships"`
+	NextPageToken string `json:"nextPageToken"`
+}
+
+func (s *SocialGoogle) retrieveGroups(ctx context.Context, client *http.Client, userData *googleUserData) ([]string, error) {
+	contains := func(slice []string, item string) bool {
+		for _, s := range slice {
+			if s == item {
+				return true
+			}
+		}
+		return false
+	}
+
+	s.log.Debug("Retrieving groups", "scopes", s.SocialBase.Config.Scopes)
+	if !contains(s.Scopes, "https://www.googleapis.com/auth/cloud-identity.groups.readonly") {
+		return nil, nil
+	}
+
+	groups := []string{}
+
+	url := fmt.Sprintf("%s?query=member_key_id=='%s'", googleIAMGroupsEndpoint, userData.Email)
```

Comment:
Because Email has been provided by Google I don't believe it could be manipulated to inject unexpected characters in the url